### PR TITLE
Changes Needed to Generate Sync RequestAuthenticator

### PR DESCRIPTION
**Pull request by AnaCoda**
[US154650](https://rally1.rallydev.com/#/?detail=/userstory/706189919605&fdp=true): Async to Sync Generator: Generate RequestAuthenticator
Added [GenerateSync] as needed, partial classes, and small changes (made Task<ID2LPrincipal> IRequestAuthenticator.AuthenticateAsync() asynchronous)

### CODE REVIEW

File: src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs [hunk 1]
```diff
@@ -22,11 +25,15 @@ Uri jwkEndpoint
 			m_jwkEndpoint = jwkEndpoint;
 		}
 
+		[GenerateSync]
 		async Task<JsonWebKeySet> IJwksProvider.RequestJwksAsync() {
 			try {
 				using( HttpResponseMessage response = await m_httpClient.GetAsync( m_jwksEndpoint ).ConfigureAwait( false ) ) {
 					response.EnsureSuccessStatusCode();
-					string jsonResponse = await response.Content.ReadAsStringAsync().ConfigureAwait( false );
+					var reader = new StreamReader(await response.Content.ReadAsStreamAsync().ConfigureAwait(false));
```
Comment (j3parker) [inline src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs:33]:
```suggestion
					using var reader = new StreamReader(await response.Content.ReadAsStreamAsync().ConfigureAwait(false));
```

(ditto elsewhere)

File: src/D2L.Security.OAuth2/Keys/Development/InMemoryPublicKeyDataProvider.cs [hunk 2]
```diff
@@ -14,30 +15,34 @@ namespace D2L.Security.OAuth2.Keys.Development {
...
-			return Task.CompletedTask;
+			await Task.CompletedTask.ConfigureAwait(false);
 		}
 
-		Task IPublicKeyDataProvider.DeleteAsync( Guid id ) {
+		[GenerateSync]
+		async Task IPublicKeyDataProvider.DeleteAsync( Guid id ) {
 			m_keys.TryRemove( id, out JsonWebKey removedKey );
-			return Task.CompletedTask;
+			await Task.CompletedTask.ConfigureAwait(false);
```
Comment (j3parker) [inline src/D2L.Security.OAuth2/Keys/Development/InMemoryPublicKeyDataProvider.cs:45]: Hm, we could transform `return Task.CompletedTask;` into just `return;` I think.

[Check out the decompilation of](https://sharplab.io/#v2:EYLgtghglgdgPgAQAwAIEEYCsBuAsAKAIQGY10A2NAJhQGEUBvAlFtUjShADjUoDEA9gIAUASkbNWUhAE5eAOloCwABwA2AUwAuGgCYJyigTABmUAOYBXAE4aAggHdoW4ShMQ1AZw0pRefFIAvgSBQA=) a method with `await Task.CompletedTask.ConfigureAwait( false );`. It's ultimately not really doing much else but it's definitely more junk than just `return Task.CompletedTask`. Mostly though it's just "weird" to see `await Task.CompletedTask` :) I'd like for `[GenerateSync]` to allow us to write idiomatic code.

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+			=> m_httpClient.SendAsync(request, completionOption);
+
+		public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
+			=> m_httpClient.SendAsync(request, completionOption, cancellationToken);
+
+#pragma warning disable D2L0018 // Avoid using dangerous methods
+		public HttpResponseMessage Get(string requestUri, HttpCompletionOption completionOption, CancellationToken cancellationToken) {
+			var task = Task.Run(() => m_httpClient.GetAsync(requestUri, completionOption, cancellationToken));
+			task.Wait();
+			return task.Result;
```
Comment (j3parker) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:60]: I'd be fine with just

```cs
=> Task.Run( () => m_httpClient.getAsync( requestUri, completionOption, cancellationToken ) ).Result;
```

but no big deal

File: src/D2L.Security.OAuth2/Validation/AccessTokens/AccessTokenValidatorFactory.cs [hunk 4]
```diff
@@ -37,7 +38,7 @@ IPublicKeyDataProvider publicKeyDataProvider
 		/// <param name="jwkEndpoint">The full URI of the remote JWK path</param>
 		/// <returns>A new <see cref="IAccessTokenValidator"/></returns>
 		public static IAccessTokenValidator CreateRemoteValidator(
-			HttpClient httpClient,
+			D2LHttpClient httpClient,
```
Comment (j3parker) [inline src/D2L.Security.OAuth2/Validation/AccessTokens/AccessTokenValidatorFactory.cs:41]: I think the public methods should take `ID2LHttpClient`?

File: src/D2L.Security.OAuth2/Keys/Development/InMemoryPublicKeyDataProvider.cs (hunk 2 above)
Comment (j3parker) [reply src/D2L.Security.OAuth2/Keys/Development/InMemoryPublicKeyDataProvider.cs:45]: I wouldn't add the `async` modifier to them, e.g. I'd hope that

```cs
Task IPublicKeyDataProvider.DeleteAsync( Guid id ) {
	m_keys.TryRemove( id, out JsonWebKey removedKey );
	return Task.CompletedTask;
}
```

could get transformed into

```cs
void IPublicKeyDataProvider.Delete( Guid id ) {
	m_keys.TryRemove( id, out JsonWebKey removedKey );
	return;
}
```

or something like that

File: src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs [hunk 5]
```diff
@@ -22,11 +25,12 @@ Uri jwkEndpoint
 			m_jwkEndpoint = jwkEndpoint;
 		}
 
+		[GenerateSync]
 		async Task<JsonWebKeySet> IJwksProvider.RequestJwksAsync() {
 			try {
 				using( HttpResponseMessage response = await m_httpClient.GetAsync( m_jwksEndpoint ).ConfigureAwait( false ) ) {
 					response.EnsureSuccessStatusCode();
-					string jsonResponse = await response.Content.ReadAsStringAsync().ConfigureAwait( false );
+					string jsonResponse = await response.Content.ReadAsStringAsync().ConfigureAwait(false);
```
Comment (j3parker) [inline src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs:33]:
```suggestion
					string jsonResponse = await response.Content.ReadAsStringAsync().ConfigureAwait( false );
```

(nit, mostly to undo the change)

**Review by j3parker — APPROVED** [review]

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs [hunk 6]
```diff
@@ -0,0 +1,125 @@
+﻿using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace D2L.Security.OAuth2.Utilities
+{
+	public class D2LHttpClient : ID2LHttpClient
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: Didn't our strategy pivot to sticking with `HttpClient` and generating the `Task.Run` wrapper?

Comment (AnaCoda) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: Correct me if I'm wrong @j3parker but I think this is a form of future-proofing so that we can implement true "sync" versions of HttpClient methods in the future. We do have the ability to wrap them in Task.Run but saved that for one-offs and not the HttpClient class

Comment (j3parker) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: I think the interface we're going to need to wrap HTTP is going to look a lot different than what we've got here (like methods like `T GetAsJson<T>( string url );` etc., high level stuff).

I'm OK if we can drop D2LHttpClient for now and rely on the generator, or if that's a hassle I don't care if we have this (just it'll change drastically in a bit, that's all.)

Comment (AnaCoda) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: It is a bit of a hassle, because given the common names "Get" and "Send", I would have to check the types of whatever is calling them. Anyway, it would have to be a new package so if things are going to change drastically anyway I feel that we shouldn't make additional changes to the generator only to revert them later

**Comment by AnaCoda** [comment]
> Very cool.
> 
> Could we try `[GenerateSync]` on one of the tests for fun? If it works to generate a sync test that would be pretty cool, heh. If it doesn't that's fine for now but I'd be curious about why.

Any tests for these changes will be in https://github.com/Brightspace/D2L.Security.OAuth2/pull/316/ (force pushed to remove them from this branch)

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs (hunk 6 above)
Comment (omsmith) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: I guess I'm unsure why we added the `ReadAsStringAsync` wrapping into the generator, when it was directly related to this HttpClient usage - and I thought we were ultimately adding it to avoid having to wrap HttpClient (at the moment, at least).

Whatever works if we're just working incrementally though. I'd rather us not cut a release of D2L.Security.OAuth2 as-is though.

Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]:
```suggestion
	public sealed class D2LHttpClient : ID2LHttpClient
```

Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]:
```suggestion
	public sealed class D2LHttpClient : ID2LHttpClient
```

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+namespace D2L.Security.OAuth2.Utilities
+{
+	public class D2LHttpClient : ID2LHttpClient
+	{
+		/// <summary>
+		/// This class implements sync versions of HTTPClient methods.
+		/// TODO: Replace the sync methods which are currently using Task.Run(async verison); task.Wait();
+		/// With (probably) HttpWebRequest
+		/// </summary>
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:14]: I think this comment belongs more on the `D2LHttpClient` class, rather than `m_httpClient`?

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs (hunk 6 above)
Comment (AnaCoda) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:8]: I see, it's because ReadAsStringAsync is part of HttpContent not HttpClient which complicates things

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+
+		public D2LHttpClient() {
+			m_httpClient = new HttpClient();
+		}
+		public D2LHttpClient(HttpMessageHandler handler) {
+			m_httpClient = new HttpClient(handler);
+		}
+		public D2LHttpClient(HttpMessageHandler handler, bool disposeHandler) { 
+			m_httpClient = new HttpClient(handler, disposeHandler);
+		}
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:25]: For the moment I would suggest making `ID2LHttpClient` and `D2LHttpClient` `internal`, using it to wrap the passed in `HttpClient`, and avoid changing the public interface of this library (as far as the factories go and whatnot (i.e. undo changes to `AccessTokenProviderFactory`)).

```suggestion
		public D2LHttpClient( HttpClient httpClient ) {
			m_httpClient = httpClient;
		}
```

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+		public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, CancellationToken cancellationToken)
+			=> m_httpClient.SendAsync(request, cancellationToken);
+
+		public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption)
+			=> m_httpClient.SendAsync(request, completionOption);
+
+		public Task<HttpResponseMessage> SendAsync(HttpRequestMessage request, HttpCompletionOption completionOption, CancellationToken cancellationToken)
+			=> m_httpClient.SendAsync(request, completionOption, cancellationToken);
+
+#pragma warning disable D2L0018 // Avoid using dangerous methods
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:56]: Would be great if we could use `[DangerousMethods.Audited( ... )] instead of disabling the `#pragma`

File: src/D2L.Security.OAuth2/Utilities/ID2LHttpClient.cs [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+using System.Threading.Tasks;
+
+namespace D2L.Security.OAuth2.Utilities
+{
+	public interface ID2LHttpClient : IDisposable
+	{
+		/// <summary>
+		/// This interface was created to allow the sync generator in D2L.CodeStyle to strip the async suffix from HttpClient calls
+		/// when it tries to convert async code to sync code.
+		/// </summary>
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Utilities/ID2LHttpClient.cs:13]: Comment belongs on `ID2LHttpClient`, not the method.

File: src/D2L.Security.OAuth2/Validation/Request/RequestAuthenticator.cs [hunk 11]
```diff
@@ -14,14 +15,16 @@ internal sealed partial class RequestAuthenticator : IRequestAuthenticator {
...
 
-		Task<ID2LPrincipal> IRequestAuthenticator.AuthenticateAsync(
+		[GenerateSync]
+		async Task<ID2LPrincipal> IRequestAuthenticator.AuthenticateAsync(
 			HttpRequestMessage request
 		) {
 			string bearerToken = request.GetBearerTokenValue();
 
-			return AuthenticateAsync( bearerToken );
+			return await AuthenticateAsync( bearerToken ).ConfigureAwait(false);
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Validation/Request/RequestAuthenticator.cs:24]: I think the latest version of the generator doesn't require this change to `async`, right?

File: Directory.Build.props [hunk 12]
```diff
@@ -24,5 +24,6 @@
 	<PropertyGroup>
 		<LangVersion>11.0</LangVersion>
 		<ManagePackageVersionsCentrally>true</ManagePackageVersionsCentrally>
+		<WarningsAsErrors>CS8785</WarningsAsErrors>
```
Comment (omsmith) [inline Directory.Build.props:27]: What is CS8785?

**Review by omsmith — COMMENTED** [review]
There's a number of cases of new code not matching our existing style. Could you get that fixed up?

* opening braces shouldn't have their own line (`class Foo : IFoo {`)
* method declarations and invocations should have spaces within the parens (`void Foo( string bar ) {`, `Foo( "Bar" );`)

File: Directory.Build.props (hunk 12 above)
Comment (AnaCoda) [reply Directory.Build.props:27]: Generator failed to generate source (want to make sure build fails if the sync generator fails). Should I comment in here?

Comment (omsmith) [reply Directory.Build.props:27]: No that's fine, just new to me and wanted to be sure it made sense. I think we'd also notice it if we manually defined the interface methods instead of using `[GenerateSync]` to also fill those out.

```csharp
public interface IRequestAuthenticator {
  Task<ID2LPrincipal> AuthenticateAsync( HttpRequestMessage request );
  ID2LPrincipal Authenticate( HttpRequestMessage request );
}

internal sealed partial class RequestAuthenticator : IRequestAuthenticator {
  [GenerateSync]
  Task<ID2LPrincipal> IRequestAuthenticator.AuthenticateAsync(
    HttpRequestMessage request
  ) { ... }
```

In the event of a failure it would complain that `RequestAuthenticator` doesn't implement the interface. Don't know if @j3parker has a preference.

Comment (j3parker) [reply Directory.Build.props:27]: Generating the sync interface for you certainly makes it easier to maintain... I dno, I don't really understand why that diagnostic isn't already an error. Like why would you be running a generator if you were cool with it crashing? lol

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs (hunk 9 above)
Comment (AnaCoda) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:45]: Is this an existing decorator or are you proposing its creation

File: Directory.Build.props (hunk 12 above)
Comment (j3parker) [reply Directory.Build.props:27]: We also insert `[Blocking]` on the generated stuff, etc. FYI

Like if we were having you specify them manually I think we'd also want to analyze them to make sure they matched what we wanted... and at that point it seems simpler to just generate.

File: src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs (hunk 9 above)
Comment (omsmith) [reply src/D2L.Security.OAuth2/Utilities/D2LHttpClient.cs:45]: Existing, just typo'd it:

https://search.d2l.dev/search?project=lms&full=DangerousMethodUsage

File: src/D2L.Security.OAuth2.TestFramework/TestAccessToken.cs [hunk 13]
```diff
@@ -6,6 +6,7 @@
 using System.Threading.Tasks;
 using D2L.Security.OAuth2.Provisioning;
 using D2L.Security.OAuth2.Scopes;
+using D2L.Security.OAuth2.Utilities;
```
Comment (omsmith) [inline src/D2L.Security.OAuth2.TestFramework/TestAccessToken.cs:9]: nit: some more no-longer-used usings left over

File: src/D2L.Security.OAuth2/Keys/Default/Data/IJwksProvider.cs [hunk 14]
```diff
@@ -1,8 +1,11 @@
-﻿using System.Threading.Tasks;
+﻿using D2L.CodeStyle.Annotations;
+using System.Threading.Tasks;
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/Data/IJwksProvider.cs:2]: nit: convention is to place System usings first, and then the rest. It should be the Visual Studio default. There's a handful of occurrences.

image

File: src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs [hunk 15]
```diff
@@ -1,14 +1,17 @@
 ﻿using System;
+using System.IO;
```
Comment (omsmith) [inline src/D2L.Security.OAuth2/Keys/Default/Data/JwksProvider.cs:2]: nit: unused

File: test/D2L.Security.OAuth2.Benchmarks/FullStackValidation/FullStackValidationBenchmark.cs [hunk 16]
```diff
@@ -12,6 +13,7 @@ namespace D2L.Security.OAuth2.Benchmarks.FullStackValidation {
 	internal abstract class FullStackValidationBenchmark : IBenchmark {
 
 		private HttpClient m_httpClient;
+		private D2LHttpClient m_D2LHttpClient;
```
Comment (omsmith) [inline test/D2L.Security.OAuth2.Benchmarks/FullStackValidation/FullStackValidationBenchmark.cs:16]: Appears unused, given it's the only addition left in the file

**Review by j3parker — APPROVED** [review]
I think we should ultimately wrap HttpClient in something that implements the sync workflows with HttpWebRequest. The interface for that wrapper would be pretty "simple", e.g. it wouldn't return `HttpResponseMessage` (that object is too complex) and tailored to our (limited needs).

If we do that we maybe don't need the `Task.Run` wrapping support in the generator, which would be ideal.

I don't think we need to hold up this PR on any of that though, and I'm keen to see some e2e stuff in the LMS.
